heft-plugins/heft-lint-plugin/src/LintPlugin.ts

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+import * as path from 'path';
+
+import { FileSystem } from '@rushstack/node-core-library';
+import type {
+  HeftConfiguration,
+  IHeftTaskSession,
+  IHeftTaskPlugin,
+  IHeftTaskRunHookOptions,
+  IScopedLogger
+} from '@rushstack/heft';
+import type {
+  IChangedFilesHookOptions,
+  IExtendedProgram,
+  IExtendedSourceFile,
+  IExtendedTypeScript,
+  ITypeScriptPluginAccessor
+} from '@rushstack/heft-typescript-plugin';
+
+import { Eslint } from './Eslint';
+import { Tslint } from './Tslint';
+import type { LinterBase } from './LinterBase';
+
+const PLUGIN_NAME: string = 'LintPlugin';
+
+export default class LintPlugin implements IHeftTaskPlugin {
+  private readonly _lintingPromises: Promise<LinterBase<unknown>[]>[] = [];
+
+  // These are initliazed by _initAsync
+  private _initPromise!: Promise<void>;
+  private _typeScript!: IExtendedTypeScript;
+  private _eslintToolPath: string | undefined;
+  private _eslintConfigFilePath: string | undefined;
+  private _tslintToolPath: string | undefined;
+  private _tslintConfigFilePath: string | undefined;
+
+  public apply(taskSession: IHeftTaskSession, heftConfiguration: HeftConfiguration): void {
+    // Use the changed files hook to kick off linting asynchronously
+    taskSession.requestAccessToPluginByName(
+      '@rushstack/heft-typescript-plugin',
+      'TypeScriptPlugin',
+      (accessor: ITypeScriptPluginAccessor) => {
+        // Hook into the changed files hook to kick off linting, which will be awaited in the run hook
+        accessor.onChangedFilesHook?.tap(PLUGIN_NAME, (changedFilesHookOptions: IChangedFilesHookOptions) => {
+          this._lintingPromises.push(
+            this._lintAsync(
+              taskSession,
+              heftConfiguration,
+              changedFilesHookOptions.program,
+              changedFilesHookOptions.changedFiles
+            )
+          );
+        });
+      }
+    );
+
+    taskSession.hooks.run.tapPromise(PLUGIN_NAME, async (options: IHeftTaskRunHookOptions) => {
+      // Run init again. Since we are intentionally swallowing init errors in the linting promises,
+      // we will await the init here in order to ensure that init errors bubble up.
```

Comment:
Maybe include something explicit in the comment here saying:

```
// `init()` will be called by a hook from the TypeScript plugin, during the TypeScript compilation. In
// order to not report linter config errors, we swallow issues with init during that hook's tap.
// Running `init()` again here without swallowing the errors allows the errors to be surfaced during the
// lint plugin's execution instead of during the TypeScript plugin's execution.
```